core/iwasm/common/wasm_shared_memory.c

```diff
@@ -130,13 +139,13 @@ notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
         || !(args.addr = wasm_runtime_malloc((uint32)total_elem_count_size))) {
         LOG_ERROR(
             "failed to allocate memory for list of atomic wait addresses");
-        os_mutex_unlock(&shared_memory_list_lock);
+        os_mutex_unlock(&wait_map_lock);
         return;
     }
 
     /* set values in list of addresses */
     bh_hash_map_traverse(wait_map, create_list_of_waiter_addresses, &args);
-    os_mutex_unlock(&shared_memory_list_lock);
+    os_mutex_unlock(&wait_map_lock);
```

Comment:
Is it better to also lock the notify process? Or, move this line to be after L153?

Reply:
It goes into deadlock since then `wasm_runtime_atomic_notify()` calls `acquire_wait_info()` that in turn tries to get `wait_map_lock`